evap/staff/importers.py

```diff
@@ -417,6 +464,83 @@ def process(cls, excel_content, test_run):
+            msg += create_user_list_string_for_message(already_related)
+
+        # since the user profiles are not necessarily saved to the database, they are not guaranteed to have a pk yet which
+        # makes anything relying on hashes unusable here (for a faster list difference)
+        users_to_add = [user for user in user_list if user not in already_related]
+
+        if test_run:
+            msg = _("{} participants would be added to the course {}:").format(len(users_to_add), course.name)
+            msg += create_user_list_string_for_message(users_to_add)
+        else:
+            course.participants.add(*users_to_add)
+            msg = _("{} Participants added to course {}").format(len(users_to_add), course.name)
+            msg += create_user_list_string_for_message(users_to_add)
+
+        self.success_messages.append(mark_safe(msg))
+
+    def process_contributors(self, course, test_run, file_content=None, source_course=None):
+        if file_content:
+            user_list, self.success_messages, self.warnings, self.errors = UserImporter.process(file_content, test_run)
+        elif source_course:
+            user_list = list(UserProfile.objects.filter(contributions__course=source_course))
+
+        already_related = Contribution.objects.filter(course=course, contributor__in=user_list).all()
+        if already_related:
+            msg = _("The following {} user(s) are already contributing to course {}:").format(len(already_related), course.name)
+            msg += create_user_list_string_for_message(already_related)
+
+        # since the user profiles are not necessarily saved to the database, they are not guaranteed to have a pk yet which
+        # makes anything relying on hashes unusable here (for a faster list difference)
+        users_to_add = [user for user in user_list if user not in already_related]
+
+        if test_run:
+            msg = _("{} contributors would be added to the course {}:").format(len(users_to_add), course.name)
+            msg += create_user_list_string_for_message(users_to_add)
+        else:
+            for user in users_to_add:
+                order = Contribution.objects.filter(course=course).count()
+                contribution = Contribution(course=course, contributor=user, order=order)
+                contribution.save()
+                course.contributions.add(contribution)
+            msg = _("{} Contributors added to course {}").format(len(users_to_add), course.name)
+            msg += create_user_list_string_for_message(users_to_add)
+
+        self.success_messages.append(mark_safe(msg))
+
+    @classmethod
+    def process(cls, import_type, course, test_run, file_content=None, source_course=None):
+        importer = cls()
+
+        if file_content is None and source_course is None:
+            raise ValueError("Expected one of the two positional arguments: file_content or source_course")
```

Comment:
actually.... it bothers me a bit that the caller knows which of the parameters is used, but we discard that knowledge and check for it again. how about two methods, `process_file_content` and `process_source_course`? each one would take only one of those two arguments. they would have some duplication, but this check here becomes unnecessary and the if/elif at the beginning of the other two process methods can be removed, they would then only get a user_list parameter. (and btw this would also fix the landscape issue ;) )

you can try that out if you want, but i'm not sure myself whether it's an improvement.